lib/pf/Authentication/Source/LDAPSource.pm

```diff
@@ -226,169 +228,194 @@ sub match {
+}
+
+sub match_first_rule {
+    my ($self, $object, $params) = @_;
+    my %common_attributes_lookup = map { $_->{value} => undef } @{$self->common_attributes()};
+    foreach my $rule ( @{$self->{'rules'}} ) {
+        next if ( (defined($params->{'rule_class'})) && ($params->{'rule_class'} ne $rule->{'class'}) );
+        if ($self->match_rule($rule, $object, $params, \%common_attributes_lookup)) {
+            return $rule;
         }
     }
-    return $self->cache_match;
+    return undef;
 }
 
+sub match_rule {
+    my ($self, $rule, $object, $params, $lookup) = @_;
+    my $conditions = $rule->{conditions};
+    #If this is a catch all return true
+    return 1 if @$conditions == 0;
+    #Split the conditions into common, ldap and member conditions
+    my ($common_conditions, $ldap_conditions, $member_conditions) = part {
+        exists $lookup->{$_->attribute} ? 0
+        : $_->operator ne $Conditions::IS_MEMBER ? 1
+        : 2
+    } @$conditions;
+    $common_conditions //= [];
+    $ldap_conditions //= [];
+    $member_conditions //= [];
+    if ($rule->match eq $Rules::ANY) {
+        return any { $self->match_condition($_, $params) } @$common_conditions
+               || any { $self->match_ldap_condition($_, $object) } @$ldap_conditions
+               || $self->match_member_conditions($object, $member_conditions, $rule->match);
+    }
+    return (@$common_conditions == 0 || all { $self->match_condition($_, $params) } @$common_conditions)
+           && (@$ldap_conditions == 0 || all { $self->match_ldap_condition($_, $object) } @$ldap_conditions)
+           && (@$member_conditions == 0 || $self->match_member_conditions($object, $member_conditions, $rule->match));
+}
 
-=head2 match_in_subclass
+sub match_member_conditions {
+    my ($self, $object, $member_conditions, $match) = @_;
+    my $dn_search = escape_filter_value($object->dn);
+    my $cached_connection = $self->_cached_connection;
+    return undef if !defined $cached_connection;
+    (my $connection, undef, undef) = @$cached_connection;
+    if (!defined($connection)) {
+      return 0;
+    }
+    if ($match eq $Rules::ANY ) {
+        return any { $self->match_group_filter($connection, $object, $dn_search, $_) } @$member_conditions;
+    }
+    return all { $self->match_group_filter($connection, $object, $dn_search, $_) } @$member_conditions;
+}
 
-C<$params> are the parameters gathered at authentication (username, SSID, connection type, etc).
+sub match_group_filter {
+    my ($self, $connection, $entry, $dn_search, $condition) = @_;
+    my $value = escape_filter_value($condition->{'value'});
+    my $attribute = escape_filter_value($entry->get_value($condition->{'attribute'}));
+    # Search for any type of group definition:
+    # - groupOfNames       => member (dn)
+    # - groupOfUniqueNames => uniqueMember (dn)
+    # - posixGroup         => memberUid (uid)
+    my $filter = "(|(member=${dn_search})(uniqueMember=${dn_search})(memberUid=${attribute}))";
```

Comment:
Exactly, and I'm affraid of other surprises like that.

As we discussed together, I expect the perf gain from this PR will be minimal since the vast majority of the auth rules are group based.

We'll see what Mr. Munro thinks about it but from our discussions, me and James believe that this is more trouble than anything else, especially when taking in consideration the small amount of cases where the object can actually be used for matching.